### PR TITLE
matrix::select_k: extra tests and benchmarks

### PR DESCRIPTION
**Pull request by achirkin**
Add a few extra test and benchmark cases; in particular:
  1. Allow specifying non-trivial input indices
  2. Allow filling the input data with infinities to see how algorithms perform in edge cases

These tests are borrowed from the controversial workaround https://github.com/rapidsai/raft/pull/1742

### CODE REVIEW

File: cpp/test/matrix/select_k.cuh [hunk 1]
```diff
@@ -223,32 +227,62 @@ struct SelectK  // NOLINT
...
       return false;
     };
     ASSERT_TRUE(hostVecMatch(ref.get_out_ids(), res.get_out_ids(), compare_ids));
   }
+
+  auto forgive_algo(const std::optional<select::Algo>& algo, IdxT ix) const -> bool
+  {
+    if (!algo.has_value()) { return false; }
+    switch (algo.value()) {
+      // not sure which algo this is.
```
Comment (tfeher) [inline cpp/test/matrix/select_k.cuh:262]: What does this mean: "not clear which one is selected because it might depend on input parameters"? or something else?

File: cpp/test/matrix/select_k.cuh [hunk 2]
```diff
@@ -223,32 +227,62 @@ struct SelectK  // NOLINT
...
+    // due to non-deterministic nature of some implementations.
+    auto compare_ids = [this](const IdxT& i, const IdxT& j) {
       if (i == j) return true;
+      auto& in_ids   = ref.get_in_ids();
+      auto& in_dists = ref.get_in_dists();
       auto ix_i = static_cast<int64_t>(std::find(in_ids.begin(), in_ids.end(), i) - in_ids.begin());
       auto ix_j = static_cast<int64_t>(std::find(in_ids.begin(), in_ids.end(), j) - in_ids.begin());
-      if (static_cast<size_t>(ix_i) >= in_ids.size() || static_cast<size_t>(ix_j) >= in_ids.size())
-        return false;
+      auto forgive_i = forgive_algo(ref.algo, i);
```
Comment (tfeher) [inline cpp/test/matrix/select_k.cuh:238]: Could you add a comment on the forgiveness logic here?

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks Artem for improving the tests. Looks good, just two small questions.

File: cpp/test/matrix/select_k.cuh (hunk 2 above)
Comment (achirkin) [reply cpp/test/matrix/select_k.cuh:238]: I explain it in more detail in an issue: https://github.com/rapidsai/raft/issues/1822

File: cpp/test/matrix/select_k.cuh (hunk 1 above)
Comment (achirkin) [reply cpp/test/matrix/select_k.cuh:262]: Yes, we don't have access to the algo selection logic here, thus we don't know what behavior to expect.

File: cpp/test/matrix/select_k.cuh (hunk 2 above)
Comment (tfeher) [reply cpp/test/matrix/select_k.cuh:238]: Thanks the issue helps, would be great to link in the comment below.

**Review by tfeher — APPROVED** [review]
Thanks @achirkin for answering my questions. LGTM.

File: cpp/test/matrix/select_k.cuh [hunk 3]
```diff
@@ -223,32 +227,62 @@ struct SelectK  // NOLINT
...
+      auto forgive_j = forgive_algo(res.algo, j);
+      // Some algorithms return invalid indices in special cases.
+      // This can be considered as TODO for us to fix.
```
Comment (achirkin) [inline cpp/test/matrix/select_k.cuh:241]:
```suggestion
      // TODO: https://github.com/rapidsai/raft/issues/1822
```

File: cpp/test/matrix/select_k.cuh (hunk 2 above)
Comment (achirkin) [reply cpp/test/matrix/select_k.cuh:238]: Thanks, good point, I added the link.
